doc/source/optimization.md

```diff
@@ -1,67 +1,264 @@
+- Enable the continuous image puller, to prepare added nodes for arriving users
+- Enable pod priority and add user placeholders, to scale up nodes ahead of real user arrivals
+- Enable the user scheduler, to pack users tight on some nodes and let others become empty and scaled down
+- Use a node pool dedicated to user pods, to avoid getting other pods on them that blocks scale down
 
-If a user is scheduled on a node requesting an image that isn't available, the
-user will have to wait. If the image is large, the wait can be 5 to 10 minutes.
-This can happen for two reasons.
-
-1. A new image is introduced during a `helm upgrade`
-
-With the `hook-image-puller` puller enabled, images will be pulled to the nodes
-before the actual `helm upgrade` thanks to a [Helm
-hook](https://docs.helm.sh/developing_charts/#hooks).
+A reasonable final configuration could look something like this:
 
 ```yaml
+scheduling:
+  userScheduler:
+    enabled: true
+  podPriority:
+    enabled: true
+  userPlaceholder:
+    enabled: true
+    replicas: 4
+  userPods:
+    nodeAffinity:
+      matchNodePurpose: require
+
 prePuller:
-  hook:
-    enabled: false
+  continuous:
+    enabled: true
+
+cull:
+  enabled: true
+  timeout: 7200
+  every: 300
 ```
 
-2. A new node was added
+## Pulling images before user arrives
 
-CONTINUE FROM HERE ... Cluster size can change through manual addition of nodes or autoscaling. When a
-new node is added to the cluster, the new node does not yet have the user image.
-A user using this new node would be forced to wait while the image is pulled
-from scratch. Ideally, it would be helpful to pre-pull images when the new node
-is added to the cluster.
+If a user pod is scheduled on a node requesting a Docker image that isn't
+available, the user will have to wait for it. If the image is large, the wait
+can be 5 to 10 minutes. This commonly occur in two situations:
 
-With the **continuous pre-puller** enabled (disabled by default), the user's
-container image will be pre-pulled when a new node is added. New nodes can for
-example be added manually or by a cluster autoscaler. The **continuous
-pre-puller** uses a
-[daemonset](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/)
-to force Kubernetes to pull the user image on all nodes as soon as a node is
-present. The continuous pre-puller uses minimal resources on all nodes and
-greatly speeds up the user pod start time.
+1. A new image is introduced (`helm upgrade`)
 
-The continuous pre-puller is disabled by default. To enable it, use the
-following snippet in your `config.yaml`:
+    With the *hook-image-puller* enabled (the default), the user images being
+    introduced will be pulled to the nodes before hub pod is updated to utilize
+    the new image. The name hook-image-puller is a technical name referring to
+    how a [Helm hook](https://docs.helm.sh/developing_charts/#hooks) is used to
+    accomplish this, a more informative name would had been
```

Comment:
```suggestion
     accomplish this, a more informative name would have been
```

I really like giving a more intuitive example name for variables that are not intuitively-named :-)